File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/GlTexture.java
package com.darkraha.opengldemoj.gl;

import org.lwjgl.stb.STBImage;
import org.lwjgl.system.MemoryStack;

import java.io.File;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;

import static org.lwjgl.opengl.GL33.*;
import static org.lwjgl.stb.STBImage.stbi_failure_reason;
import static org.lwjgl.stb.STBImage.stbi_image_free;

public class GlTexture {
    public final String name;
    public final int idTexture;
    public final int textureUnit;
    public final int textureType;



    public GlTexture(int idTexture, String name) {
        this(idTexture, GL_TEXTURE0, GL_TEXTURE_2D, name);
    }

    public GlTexture(int idTexture, int textureUnitDefault, int textureType, String name) {
        this.idTexture = idTexture;
        this.name = name;
        this.textureUnit = textureUnitDefault;
        this.textureType = textureType;
    }


    public void dispose() {
        glBindTexture(textureType,0);
        glDeleteTextures(idTexture);
    }


    public static GlTexture newTexture2D(String resPath, String name) {
        return new TextureBuilder(resPath, name).build2D();
    }


    public static GlTexture newSolidColorTexture(int r, int g, int b, String name) {
        int level = 0;
        int width = 1;
        int height = 1;
        int border = 0;
        int idTexture = glGenTextures();

        glBindTexture(GL_TEXTURE_2D, idTexture);

        try (MemoryStack stack = MemoryStack.stackPush()) {
            ByteBuffer buffer = stack.bytes((byte) r, (byte) g, (byte) b, (byte) 0xff);
            glTexImage2D(GL_TEXTURE_2D, level, GL_RGBA, width, height, border,
                    GL_RGBA, GL_UNSIGNED_BYTE, buffer);
        }

        glBindTexture(GL_TEXTURE_2D, 0);
        return new GlTexture(idTexture, GL_TEXTURE0, GL_TEXTURE_2D, name);
    }


    public static void loadTexture(String resPath, TextureBuilder tbuilder) {
        try (MemoryStack stack = MemoryStack.stackPush()) {

            IntBuffer w = stack.mallocInt(1);
            IntBuffer h = stack.mallocInt(1);
            IntBuffer channels = stack.mallocInt(1);
            URL resource = Thread.currentThread().getClass().getResource(resPath);
            String filePath = new File(resource.toURI()).getAbsolutePath(); // encode spaces
            ByteBuffer buffer = STBImage.stbi_load(filePath, w, h, channels, 4);

            if (buffer == null) {
                throw new Exception("Can't load file " + filePath + " " + stbi_failure_reason());
            }

            tbuilder.mImageData = buffer;
            tbuilder.mWidth = w.get();
            tbuilder.mHeight = h.get();
            tbuilder.mFormat = GL_RGBA;

            // stbi_image_free(buffer);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    public static class TextureBuilder {
        String mName;
        int mIdTexture;
        int mTextureUnit = GL_TEXTURE0;
        int mTextureType;

        String mResPath;
        int mBorder;
        int mWidth;
        int mHeight;
        int mFormat = GL_RGBA;
        ByteBuffer mImageData;

        boolean mGenMipmap = true;
        boolean mClampToEdge = false;

        public TextureBuilder(String resPath, String name) {
            mName = name;
            mResPath = resPath;
        }


        public TextureBuilder genMipmap(boolean v) {
            mGenMipmap = v;
            return this;
        }

        public TextureBuilder clampToEdge(boolean v) {
            mClampToEdge = v;
            return this;
        }

        public GlTexture build2D() {
            mTextureType = GL_TEXTURE_2D;
            mIdTexture = glGenTextures();
            glBindTexture(mTextureType, mIdTexture);
            loadTexture(mResPath, this);
            glTexImage2D(
                    mTextureType, 0, GL_RGBA, mWidth, mHeight, 0,
                    GL_RGBA, GL_UNSIGNED_BYTE, mImageData
            );

            glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
            glTexParameteri(mTextureType, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

            if (mGenMipmap) {
                glGenerateMipmap(mTextureType);
                glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            }

            if (mClampToEdge) {
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
            }

            stbi_image_free(mImageData);
            glBindTexture(GL_TEXTURE_2D, 0);
            return new GlTexture(mIdTexture, mTextureUnit, mTextureType, mName);
        }

    }

}

File: lwjglkt/src/main/kotlin/com/darkraha/opengldemokt/gl/GlTexture.kt
package com.darkraha.opengldemokt.gl


import org.lwjgl.opengl.GL33.*

import org.lwjgl.stb.STBImage
import org.lwjgl.system.MemoryStack
import java.io.File
import java.lang.Exception
import java.nio.ByteBuffer

class GlTexture(val idTexture: Int, val textureUnit: Int, val textureType: Int, val name: String) {
    constructor(idTexture: Int, name: String) : this(idTexture, GL_TEXTURE0, GL_TEXTURE_2D, name) {}

    fun dispose() {
        glBindTexture(textureType, 0)
        glDeleteTextures(idTexture)
    }

    class TextureBuilder(var mResPath: String, var mName: String) {
        var mIdTexture = 0
        var mTextureUnit = GL_TEXTURE0
        var mTextureType = 0
        var mBorder = 0
        var mWidth = 0
        var mHeight = 0
        var mFormat = GL_RGBA
        var mImageData: ByteBuffer? = null
        var mGenMipmap = true
        var mClampToEdge = false
        fun genMipmap(v: Boolean): TextureBuilder {
            mGenMipmap = v
            return this
        }

        fun clampToEdge(v: Boolean): TextureBuilder {
            mClampToEdge = v
            return this
        }

        fun build2D(): GlTexture {
            mTextureType = GL_TEXTURE_2D
            mIdTexture = glGenTextures()
            glBindTexture(mTextureType, mIdTexture)
            loadTexture(mResPath, this)

            glTexImage2D(
                mTextureType, 0, GL_RGBA, mWidth, mHeight, 0,
                GL_RGBA, GL_UNSIGNED_BYTE, mImageData
            )

            glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
            glTexParameteri(mTextureType, GL_TEXTURE_MAG_FILTER, GL_LINEAR)

            if (mGenMipmap) {
                glGenerateMipmap(mTextureType)
                glTexParameteri(mTextureType, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR)
            }
            if (mClampToEdge) {
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE)
                glTexParameteri(mTextureType, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE)
            }
            STBImage.stbi_image_free(mImageData)
            glBindTexture(GL_TEXTURE_2D, 0)
            return GlTexture(mIdTexture, mTextureUnit, mTextureType, mName)
        }
    }

    companion object {
        fun newTexture2D(resPath: String, name: String): GlTexture {
            return TextureBuilder(resPath, name).build2D()
        }

        fun newSolidColorTexture( r: Int, g: Int, b: Int, name: String): GlTexture {
            val level = 0
            val width = 1
            val height = 1
            val border = 0
            val idTexture = glGenTextures()
            glBindTexture(GL_TEXTURE_2D, idTexture)
            MemoryStack.stackPush().use { stack ->
                val buffer = stack.bytes(
                    r.toByte(), g.toByte(), b.toByte(),
                    0xff.toByte()
                )
                glTexImage2D(
                    GL_TEXTURE_2D, level, GL_RGBA, width, height, border,
                    GL_RGBA, GL_UNSIGNED_BYTE, buffer
                )
            }
            glBindTexture(GL_TEXTURE_2D, 0)
            return GlTexture(idTexture, GL_TEXTURE0, GL_TEXTURE_2D, name)
        }

        fun loadTexture(resPath: String?, tbuilder: TextureBuilder) {
            try {
                MemoryStack.stackPush().use { stack ->
                    val w = stack.mallocInt(1)
                    val h = stack.mallocInt(1)
                    val channels = stack.mallocInt(1)
                    val resource = Thread.currentThread().javaClass.getResource(resPath)
                    val filePath = File(resource.toURI()).absolutePath // encode spaces
                    val buffer = STBImage.stbi_load(filePath, w, h, channels, 4)
                        ?: throw Exception("Can't load file " + filePath + " " + STBImage.stbi_failure_reason())
                    tbuilder.mImageData = buffer
                    tbuilder.mWidth = w.get()
                    tbuilder.mHeight = h.get()
                    tbuilder.mFormat = GL_RGBA
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}